File: src/main/intellibitz/sted/ui/MappingTableModel.java
package intellibitz.sted.ui;

import intellibitz.sted.event.IMessageEventSource;
import intellibitz.sted.event.IMessageListener;
import intellibitz.sted.event.MessageEvent;
import intellibitz.sted.fontmap.FontMap;
import intellibitz.sted.fontmap.FontMapEntries;
import intellibitz.sted.fontmap.FontMapEntry;
import intellibitz.sted.util.Resources;

import javax.swing.table.AbstractTableModel;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MappingTableModel
        extends AbstractTableModel
        implements IMessageEventSource {
    private static final String[] names = new String[]
            {
                    Resources.getResource(Resources.TITLE_TABLE_COLUMN_SYMBOL1),
                    Resources.getResource(Resources.TITLE_TABLE_COLUMN_EQUALS),
                    Resources.getResource(Resources.TITLE_TABLE_COLUMN_SYMBOL2),
                    Resources.getResource(
                            Resources.TITLE_TABLE_COLUMN_FIRST_LETTER),
                    Resources.getResource(
                            Resources.TITLE_TABLE_COLUMN_LAST_LETTER),
                    Resources.getResource(
                            Resources.TITLE_TABLE_COLUMN_FOLLOWED_BY),
                    Resources.getResource(
                            Resources.TITLE_TABLE_COLUMN_PRECEDED_BY)
            };
    private FontMap fontMap;
    private FontMapEntries fontMapEntries;
    private List<FontMapEntry> entries;
    private IMessageListener messageListener;
    private MessageEvent messageEvent;

    public MappingTableModel() {
        super();
        messageEvent = new MessageEvent(this);
    }

    public void setFontMap(FontMap fontMap) {
        this.fontMap = fontMap;
        fontMapEntries = fontMap.getEntries();
        entries = new ArrayList<FontMapEntry>(fontMapEntries.values());
        Collections.sort(entries, Collections.reverseOrder());
        super.fireTableDataChanged();
    }

    public int getRowCount() {
        if (fontMapEntries == null) {
            return 0;
        }
        return fontMapEntries.size();
    }


    public Object getValueAt(int row, int col) {
        if (col == 1) {
            return Resources.EQUALS;
        }
        if (entries != null && !entries.isEmpty()) {
            final FontMapEntry entry = (FontMapEntry) entries.get(row);
            if (col == 0) {
                return entry.getFrom();
            } else if (col == 2) {
                return entry.getTo();
            } else if (col == 3) {
                if (entry.isBeginsWith()) {
                    return Boolean.TRUE;
                }
                return Boolean.FALSE;
            } else if (col == 4) {
                if (entry.isEndsWith()) {
                    return Boolean.TRUE;
                }
                return Boolean.FALSE;
            } else if (col == 5) {
                final String val = entry.getFollowedBy();
                return val == null ? Resources.EMPTY_STRING : val;
            } else if (col == 6) {
                final String val = entry.getPrecededBy();
                return val == null ? Resources.EMPTY_STRING : val;
            }
            return Resources.EQUALS;
        }
        return Resources.EQUALS;
    }

    public FontMapEntry getValueAt(int row) {
        if (entries != null && !entries.isEmpty()) {
            return (FontMapEntry) entries.get(row);
        }
        return null;
    }

    public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
        if (aValue != null) {
            final FontMapEntry entry = (FontMapEntry) entries.get(rowIndex);
            if (entry != null) {
                switch (columnIndex) {
                    case 0:
                        if (!entry.getFrom().equals(aValue)) {
                            final FontMapEntry edited =
                                    (FontMapEntry) entry.clone();
                            edited.setFrom(aValue.toString());
                            // do a strick check - just like add
                            if (!fontMapEntries.isValid(edited)) {
                                fireMessagePosted(
                                        "Already mapped - Invalid Edit");
                                break;
                            }
                            edited.setFrom(entry.getFrom());
                            edited.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.getUndo().add(edited);
                            entry.setFrom(aValue.toString());
                            entry.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.reKey(edited, entry);
                            fontMap.setDirty(true);
                            super.fireTableCellUpdated(rowIndex, columnIndex);
                        }
                        break;
                    case 2:
                        if (!entry.getTo().equals(aValue)) {
                            final FontMapEntry edited =
                                    (FontMapEntry) entry.clone();
                            edited.setTo(aValue.toString());
                            if (!fontMapEntries.isValidEdit(edited)) {
                                fireMessagePosted(
                                        "Already mapped - Invalid Edit");
                                break;
                            }
                            edited.setTo(entry.getTo());
                            edited.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.getUndo().add(edited);
                            entry.setTo(aValue.toString());
                            entry.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.reKey(edited, entry);
                            fontMap.setDirty(true);
                            super.fireTableCellUpdated(rowIndex, columnIndex);
                        }
                        break;
                    case 3:
                        final boolean begins = (Boolean) aValue;
                        if (!entry.isBeginsWith() == begins) {
                            final FontMapEntry edited =
                                    (FontMapEntry) entry.clone();
                            edited.setBeginsWith(begins);
                            if (!fontMapEntries.isValidEdit(edited)) {
                                fireMessagePosted(
                                        "Already mapped - Invalid Edit");
                                break;
                            }
                            edited.setBeginsWith(entry.isBeginsWith());
                            edited.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.getUndo().add(edited);
                            entry.setBeginsWith(begins);
                            entry.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.reKey(edited, entry);
                            fontMap.setDirty(true);
                            super.fireTableCellUpdated(rowIndex, columnIndex);
                        }
                        break;
                    case 4:
                        final boolean ends = (Boolean) aValue;
                        if (!entry.isEndsWith() == ends) {
                            final FontMapEntry edited =
                                    (FontMapEntry) entry.clone();
                            edited.setEndsWith(ends);
                            if (!fontMapEntries.isValidEdit(edited)) {
                                fireMessagePosted(
                                        "Already mapped - Invalid Edit");
                                break;
                            }
                            edited.setEndsWith(entry.isEndsWith());
                            edited.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.getUndo().add(edited);
                            entry.setEndsWith(ends);
                            entry.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.reKey(edited, entry);
                            fontMap.setDirty(true);
                            super.fireTableCellUpdated(rowIndex, columnIndex);
                        }
                        break;
                    case 5:
                        if (!aValue.equals(entry.getFollowedBy())) {
                            if (entry.getFollowedBy() == null &&
                                    Resources.EMPTY_STRING.equals(aValue)) {
                                break;
                            }
                            final FontMapEntry edited =
                                    (FontMapEntry) entry.clone();
                            edited.setFollowedBy(aValue.toString());
                            if (!fontMapEntries.isValidEdit(edited)) {
                                fireMessagePosted(
                                        "Already mapped - Invalid Edit");
                                break;
                            }
                            edited.setFollowedBy(entry.getFollowedBy());
                            edited.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.getUndo().add(edited);
                            entry.setFollowedBy(aValue.toString());
                            entry.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.reKey(edited, entry);
                            fontMap.setDirty(true);
                            super.fireTableCellUpdated(rowIndex, columnIndex);
                        }
                        break;
                    case 6:
                        if (!aValue.equals(entry.getPrecededBy())) {
                            if (entry.getPrecededBy() == null &&
                                    Resources.EMPTY_STRING.equals(aValue)) {
                                break;
                            }
                            final FontMapEntry edited =
                                    (FontMapEntry) entry.clone();
                            edited.setPrecededBy(aValue.toString());
                            if (!fontMapEntries.isValidEdit(edited)) {
                                fireMessagePosted(
                                        "Already mapped - Invalid Edit");
                                break;
                            }
                            edited.setPrecededBy(entry.getPrecededBy());
                            edited.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.getUndo().add(edited);
                            entry.setPrecededBy(aValue.toString());
                            entry.setStatus(Resources.ENTRY_STATUS_EDIT);
                            fontMapEntries.reKey(edited, entry);
                            fontMap.setDirty(true);
                            super.fireTableCellUpdated(rowIndex, columnIndex);
                        }
                        break;
                }
            }
        }
    }

    /*
     * JTable uses this method to determine the default renderer/
     * editor for each cell.  If we didn't implement this method,
     * then the last column would contain text ("true"/"false"),
     * rather than a check box.
     */
    public Class getColumnClass(int c) {
        return getValueAt(0, c).getClass();
    }

    /*
     * Don't need to implement this method unless your table's
     * editable.
     */
    public boolean isCellEditable(int row, int col) {
        //Note that the data/cell address is constant,
        //no matter where the cell appears onscreen.
        return !(col == 1);
    }

    public int getColumnCount() {
        return names.length;
    }

    public String getColumnName(int column) {
        return names[column];
    }

    public FontMap getFontMap() {
        return fontMap;
    }

    private void fireMessagePosted(String message) {
        messageEvent.setMessage(message);
        messageListener.messagePosted(messageEvent);
    }

    public void fireMessagePosted() {
        messageListener.messagePosted(messageEvent);
    }

    public void addMessageListener(IMessageListener messageListener) {
        this.messageListener = messageListener;
    }
}

File: src/main/intellibitz/sted/ui/MappingTableModel.kt
package sted.ui

import sted.event.IMessageEventSource
import sted.event.IMessageListener
import sted.event.MessageEvent
import sted.fontmap.FontMap
import sted.fontmap.FontMapEntries
import sted.fontmap.FontMapEntry
import sted.io.Resources.getResource
import java.util.*
import javax.swing.table.AbstractTableModel

class MappingTableModel : AbstractTableModel(), IMessageEventSource {
    private val messageEvent = MessageEvent(this)
    lateinit var fontMap: FontMap
    private lateinit var fontMapEntries: FontMapEntries
    private val entries: MutableList<FontMapEntry> = ArrayList()
    private var messageListener: IMessageListener? = null
    override fun getRowCount(): Int {
        try {
            return fontMapEntries.size()
        } catch (e: UninitializedPropertyAccessException) {
            return 0
        }
    }

    override fun getValueAt(row: Int, col: Int): Any {
        if (col == 1) {
            return "="
        }
        if (entries.isNotEmpty()) {
            val entry = entries[row]
            when (col) {
                0 -> {
                    return entry.from
                }
                2 -> {
                    return entry.to
                }
                3 -> {
                    return if (entry.isBeginsWith) {
                        java.lang.Boolean.TRUE
                    } else java.lang.Boolean.FALSE
                }
                4 -> {
                    return if (entry.isEndsWith) {
                        java.lang.Boolean.TRUE
                    } else java.lang.Boolean.FALSE
                }
                5 -> {
                    val `val` = entry.followedBy
                    return `val` ?: ""
                }
                6 -> {
                    val `val` = entry.precededBy
                    return `val` ?: ""
                }
                else -> return "="
            }
        }
        return "="
    }

    fun getValueAt(row: Int): FontMapEntry? {
        return if (entries.isNotEmpty()) {
            entries[row]
        } else null
    }

    override fun setValueAt(aValue: Any, rowIndex: Int, columnIndex: Int) {
        val entry = entries[rowIndex]
        when (columnIndex) {
            0 -> if (entry.from != aValue) {
                val edited = entry.clone() as FontMapEntry
                edited.from = aValue.toString()
                // do a strick check - just like add
                if (!fontMapEntries.isValid(edited)) {
                    fireAlreadyMapped()
                    return
                }
                edited.from = entry.from
                edited.setStatus(2)
                fontMapEntries.undo.add(edited)
                entry.from = aValue.toString()
                entry.setStatus(2)
                fontMapEntries.reKey(edited, entry)
                fontMap.isDirty = true
                super.fireTableCellUpdated(rowIndex, columnIndex)
            }
            2 -> if (entry.to != aValue) {
                val edited = entry.clone() as FontMapEntry
                edited.to = aValue.toString()
                if (fontMapEntries.isValidEdit(edited)) {
                    fireAlreadyMapped()
                    return
                }
                edited.to = entry.to
                edited.setStatus(2)
                fontMapEntries.undo.add(edited)
                entry.to = aValue.toString()
                entry.setStatus(2)
                fontMapEntries.reKey(edited, entry)
                fontMap.isDirty = true
                super.fireTableCellUpdated(rowIndex, columnIndex)
            }
            3 -> {
                val begins = aValue as Boolean
                if (!entry.isBeginsWith == begins) {
                    val edited = entry.clone() as FontMapEntry
                    edited.isBeginsWith = begins
                    if (fontMapEntries.isValidEdit(edited)) {
                        fireAlreadyMapped()
                        return
                    }
                    edited.isBeginsWith = entry.isBeginsWith
                    edited.setStatus(2)
                    fontMapEntries.undo.add(edited)
                    entry.isBeginsWith = begins
                    entry.setStatus(2)
                    fontMapEntries.reKey(edited, entry)
                    fontMap.isDirty = true
                    super.fireTableCellUpdated(rowIndex, columnIndex)
                }
            }
            4 -> {
                val ends = aValue as Boolean
                if (!entry.isEndsWith == ends) {
                    val edited = entry.clone() as FontMapEntry
                    edited.isEndsWith = ends
                    if (fontMapEntries.isValidEdit(edited)) {
                        fireAlreadyMapped()
                        return
                    }
                    edited.isEndsWith = entry.isEndsWith
                    edited.setStatus(2)
                    fontMapEntries.undo.add(edited)
                    entry.isEndsWith = ends
                    entry.setStatus(2)
                    fontMapEntries.reKey(edited, entry)
                    fontMap.isDirty = true
                    super.fireTableCellUpdated(rowIndex, columnIndex)
                }
            }
            5 -> if (aValue != entry.followedBy) {
                if (entry.followedBy == null && "" == aValue) {
                    return
                }
                val edited = entry.clone() as FontMapEntry
                edited.followedBy = aValue.toString()
                if (fontMapEntries.isValidEdit(edited)) {
                    fireAlreadyMapped()
                    return
                }
                edited.followedBy = entry.followedBy
                edited.setStatus(2)
                fontMapEntries.undo.add(edited)
                entry.followedBy = aValue.toString()
                entry.setStatus(2)
                fontMapEntries.reKey(edited, entry)
                fontMap.isDirty = true
                super.fireTableCellUpdated(rowIndex, columnIndex)
            }
            6 -> if (aValue != entry.precededBy) {
                if (entry.precededBy == null && "" == aValue) {
                    return
                }
                val edited = entry.clone() as FontMapEntry
                edited.precededBy = aValue.toString()
                if (fontMapEntries.isValidEdit(edited)) {
                    fireAlreadyMapped()
                    return
                }
                edited.precededBy = entry.precededBy
                edited.setStatus(2)
                fontMapEntries.undo.add(edited)
                entry.precededBy = aValue.toString()
                entry.setStatus(2)
                fontMapEntries.reKey(edited, entry)
                fontMap.isDirty = true
                super.fireTableCellUpdated(rowIndex, columnIndex)
            }
        }
    }

    /*
     * JTable uses this method to determine the default renderer/
     * editor for each cell.  If we didn't implement this method,
     * then the last column would contain text ("true"/"false"),
     * rather than a check box.
     */
    override fun getColumnClass(c: Int): Class<*> {
        return getValueAt(0, c).javaClass
    }

    /*
     * Don't need to implement this method unless your table's
     * editable.
     */
    override fun isCellEditable(row: Int, col: Int): Boolean {
        //Note that the data/cell address is constant,
        //no matter where the cell appears onscreen.
        return col != 1
    }

    override fun getColumnCount(): Int {
        return names.size
    }

    override fun getColumnName(column: Int): String {
        return names[column] ?: ""
    }

    fun loadFontMapEntries(fontMap: FontMap) {
        this.fontMap = fontMap
        fontMapEntries = fontMap.entries
        entries.clear()
        entries.addAll(fontMapEntries.values())
        Collections.sort(entries, Collections.reverseOrder())
        super.fireTableDataChanged()
    }

    private fun fireAlreadyMapped() {
        messageEvent.message = "Already mapped - Invalid Edit"
        fireMessagePosted()
    }

    override fun fireMessagePosted() {
        messageListener?.messagePosted(messageEvent)
    }

    override fun addMessageListener(messageListener: IMessageListener) {
        this.messageListener = messageListener
    }

    companion object {
        private var names = arrayOf(
            getResource("title.table.column.symbol1"),
            getResource("title.table.column.equals"),
            getResource("title.table.column.symbol2"),
            getResource("title.table.column.letter1"),
            getResource("title.table.column.letter2"),
            getResource("title.table.column.followed"),
            getResource("title.table.column.preceded")
        )
    }
}